sig-api-machinery/design-docs/admission-control-webhooks.md

```diff
@@ -0,0 +1,934 @@
+*   In kube-up.sh and other installers, change flag per previous item. 
+*   Ensure able to monitor latency and rejection from webhooks. See [Monitorability](#monitorability).
+*   Don't send internal objects.   See [#49733](https://github.com/kubernetes/kubernetes/issues/49733)
+*   Serialize mutating Webhooks into order in the apiregistration. Leave non-mutating in parallel.
+*   Good Error Messages	.  See [Good Error Messages](#good-error-messages)	
+*   Conversion logic in GenericWebhook to send converted resource to webhook.  See [Conversion](#conversion) and [#49733](https://github.com/kubernetes/kubernetes/issues/49733).
+*   Schedule discussion around resiliency to down webhooks and bootstrapping
+
+
+# Design Discussion 
+
+
+## Why Webhooks First  
+
+We will do webhooks beta before initializers beta because:
+
+
+
+1.  **Serves Most Use Cases**: We reviewed code of all current use cases, namely: Kubernetes Built-in Admission Controllers, OpenShift Admission Controllers, Istio & Service Catalog.  (See also [Use Cases Detailed Descriptions](#use-cases-detailed-descriptions).)  All of those use cases are well served by mutating and non-mutating webhooks.  (See also [Comparison of Webhooks and Initializers](#comparison-of-webhooks-and-initializers)).
+1.  **Less Work**: An engineer quite experienced with both code bases estimated that it is less work to adding Mutating Webhooks and bring both kinds of webhooks to beta; than to bring non-mutating webhooks and initializers to Beta.  Some open issues with Initializers with long expected development time include quota replenishment bug, and controller awareness of uninitialized objects.
+1.  **API Consistency**: Prefer completing one related pair of interfaces (both kinds of webhooks) at the same time.
+
+
+## Why Support Mutation for Beta 
+
+Based on experience and feedback from the alpha phase of both Webhooks and Initializers, we believe Webhooks Beta should support mutation because:  
+
+
+
+1.  We have lots of use cases to inform this (both from Initializers, and Admission Controllers) to ensure we have needed features
+1.  We have experience with Webhooks API already to give confidence in the API.  The registration API will be quite similar except in the responses.
+1.  There is a strong community demand for something that satisfies a mutating case.
+
+
+## Plan for Existing Initializer-Clients 
+
+After the release of 1.9, we will advise users who currently use initializers to:
+
+
+
+*   Move to Webhooks if their use case fits that model well.
+*   Provide SIG-API-Machinery with feedback if Initializers is a better fit. 
+
+We will continue to support Initializers as an Alpha API in 1.9.
+
+We will point out https://github.com/caesarxuchao/example-webhook-admission-controller and extend it to support mutating webhooks, and as that Initializer docs point to it (e.g. https://github.com/kelseyhightower/kubernetes-initializer-tutorial)
```

Comment:
I think we should be less specific about exactly which example we'll doc.  I think https://github.com/openshift/generic-admission-server and https://github.com/openshift/kubernetes-namespace-reservation are probably more complete, but I don't think we need to argue about it here.

Reply:
added weasel words